sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/ForeachWriterTable.scala

```diff
@@ -141,17 +143,33 @@ class ForeachDataWriter[T](
       writer.process(rowConverter(record))
     } catch {
       case t: Throwable =>
-        writer.close(t)
+        closeWriter(t)
         throw t
     }
   }
 
   override def commit(): WriterCommitMessage = {
-    writer.close(null)
+    closeWriter(null)
     ForeachWriterCommitMessage
   }
 
-  override def abort(): Unit = {}
+  override def abort(): Unit = {
```

Comment:
It looks to me that `abort` should know the exception, so that it knows why the task failed and how to do the clean up. What do you think? @rdblue

Reply:
What is the use case for passing the exception to the writer? Shouldn't the writer always take the same action on abort, no matter what the reason to abort was?

Reply:
This issue is fixing a regression bug introduced by migration ForeachSink to DSv2, we need to pass the exception to the writer in the whole iteration process, just like the original behavior [here](https://github.com/apache/spark/pull/20951/files#diff-98acda846a9dd63efc42e0957594e05dL57). As current logic, we can only get the exception in `writer.process`.

Reply:
@xuanyuanking, what is the exception used for? I'm trying to understand why the abort would behave differently depending on the exception.

Reply:
In this case, we need to call `writer.close` in `abort`, the exception is necessary for the close API. https://github.com/apache/spark/blob/b3ffd8be14779cbb824d14b409f0a6eab93444ba/sql/core/src/main/scala/org/apache/spark/sql/ForeachWriter.scala#L129-L141

Reply:
I can see that the exception is passed to close. My question is: what does the writer do differently based on the exception? If this is to satisfy an API and you can pass any exception, then I don't think it matters. There's effort required to get the exception to pass through and without a reasonable use case I'm wondering why it is necessary to do it.

Reply:
Ah I see, thanks for your explanation Ryan.
Gives a scenario of `DataWritingSparkTask`, both dataWriter.wirte, dataWriter.commit, and Spark task self could throw different exceptions. In the close API of DataWriter, I think `CommitDeniedException` and the exception threw by `writer.commit` can deal with separately. Please correct me if I'm wrong, Thanks.

Reply:
> I think CommitDeniedException and the exception threw by writer.commit can deal with separately

Sure, but what is done differently in close? For example, is the underlying not closed in one path for some reason? I can't think of anything reasonable for close to do differently, so I'd like to know what the specific use case you're support is.

Reply:
The idea is brought up at https://github.com/apache/spark/pull/13342#discussion_r64843903 , but there is no concrete use cases mentioned at that time. cc @tdas

Reply:
If there isn't a use case, then I'm very reluctant to change the way v2 works.